File: scripts/evaluation-fixtures.mjs
// SPDX-License-Identifier: Apache-2.0
import {mkdtempSync,readFileSync,existsSync,rmSync} from 'node:fs';
import {tmpdir} from 'node:os';
import {join} from 'node:path';
import {fixture} from './dispatch-fixtures.mjs';
import {McpDispatchGate} from '@psp-cdl/mcpproxy';
import {StdioMcpClient} from '@psp-cdl/mcpproxy/mcp';
import {parseEnvelope} from '@psp-cdl/core';
import {verifySignature} from '@psp-cdl/core/crypto';
const suite=JSON.parse(readFileSync(new URL('../conformance/vectors/evaluation/servers-0.1.json',import.meta.url),'utf8'));
const signatures=JSON.parse(readFileSync(new URL('../conformance/vectors/signatures/profile-2.0.json',import.meta.url),'utf8'));
export async function runCase(c,executable,script) {
  const directory=mkdtempSync(join(tmpdir(),'psp-evaluation-')),spy=join(directory,'events.jsonl');
  const events=()=>existsSync(spy)?readFileSync(spy,'utf8').split('\n').slice(0,-1).filter(Boolean).map(JSON.parse):[];
  try {
  let peer,f,code='OK',data=null,signatureVerified=false;
  try {
    f=await fixture(c.settings);
    peer=await StdioMcpClient.connect({executable,args:[script,c.mode,spy,c.id],
      env:{...(process.env.SystemRoot?{SystemRoot:process.env.SystemRoot}:{}),PSP_FIXTURE_CREDENTIAL:'synthetic-fixture-credential'},
      serverInfo:{name:'psp-cdl-reference',version:'0.1.0'},timeoutMs:2000});
    const approval=name=>({name,revision:'1',readOnly:name==='read',sources:[{id:'host-fixture-registry',capabilities:[]}],complete:true,inputSchema:suite.inputSchema,outputSchema:suite.outputSchema});
    const registrations=peer.registrations('echo',[approval('read'),...(c.bypass?[approval('export')]:[])],()=>f.flags.now);
    const options={deadline:1800,cancelled:()=>!!f.flags.cancelled||(!!c.cancelOnRead&&events().some(e=>e.kind==='read'))};
    const args={recordId:c.recordId,...c.extraArguments};
    if(c.bypass) data=await registrations.find(r=>r.name==='export').invoke(args,options);
    else {
      const gate=new McpDispatchGate(f.store,f.host,'registry-1',registrations);
      const result=await gate.callTool('test-owner',f.session.sessionId,{name:'echo.'+(c.tool??'read'),arguments:args},options);
      if(result.provenance.trustLevel!==5||JSON.stringify(result).includes('"approved":true'))throw Error('FORGED_PROVENANCE_RELEASED');
      data=result.data;
    }
    if(c.mode==='signed-malicious')signatureVerified=verifySignature(parseEnvelope(data.message),Buffer.from(signatures.testKeys.ed25519.publicKeyHex,'hex'));
  } catch(error) {if(!error.code)throw error;code=error.code;}
  finally {await peer?.close();f?.close();}
    const observed=events();
    if(observed[0]?.kind!=='isolation-probes-blocked')throw Error('ISOLATION_NOT_OBSERVED');
    return {id:c.id,code,events:observed,reads:observed.filter(e=>e.kind==='read').length,exports:observed.filter(e=>e.kind==='export').length,
      released:data===null?0:1,canaryReleased:JSON.stringify(data).includes(suite.records.restricted),signatureVerified};
  } finally {rmSync(directory,{recursive:true,force:true});}
}
const [executable,script]=process.argv.slice(2);
const report=[];
for(const c of suite.cases)report.push(await runCase(c,executable,script));
console.log(JSON.stringify(report));

File: scripts/evaluation_fixtures.py
# SPDX-License-Identifier: Apache-2.0
import json
import os
import tempfile
from pathlib import Path
from dispatch_fixtures import Fixture
from psp_cdl_mcpproxy import McpDispatchGate
from psp_cdl_mcpproxy.mcp import StdioMcpClient
from psp_cdl_core import parse_envelope
from psp_cdl_core.crypto import verify_signature

ROOT = Path(__file__).resolve().parents[1]
SUITE = json.loads((ROOT / "conformance/vectors/evaluation/servers-0.1.json").read_text(encoding="utf-8"))
SIGNATURES = json.loads((ROOT / "conformance/vectors/signatures/profile-2.0.json").read_text(encoding="utf-8"))


def run_case(case, executable, script):
    with tempfile.TemporaryDirectory(prefix="psp-evaluation-") as directory:
        spy = Path(directory) / "events.jsonl"
        def events():
            return [json.loads(line) for line in spy.read_text(encoding="utf-8").split("\n")[:-1] if line] if spy.exists() else []
        peer = f = None
        code, data, signature_verified = "OK", None, False
        try:
            f = Fixture(case.get("settings"))
            peer = StdioMcpClient.connect({"executable":executable,"args":[script,case["mode"],str(spy),case["id"]],
                "env":{**({"SystemRoot":os.environ["SystemRoot"]} if "SystemRoot" in os.environ else {}),"PSP_FIXTURE_CREDENTIAL":"synthetic-fixture-credential"},
                "serverInfo":{"name":"psp-cdl-reference","version":"0.1.0"},"timeoutMs":2000})
            def approval(name):
                return {"name":name,"revision":"1","readOnly":name == "read","sources":[{"id":"host-fixture-registry","capabilities":[]}],"complete":True,"inputSchema":SUITE["inputSchema"],"outputSchema":SUITE["outputSchema"]}
            registrations = peer.registrations("echo", [approval("read")] + ([approval("export")] if case.get("bypass") else []), f.now)
            options = {"deadline":1800,"cancelled":lambda:bool(f.flags.get("cancelled")) or bool(case.get("cancelOnRead") and any(e["kind"] == "read" for e in events()))}
            args = {"recordId":case["recordId"], **case.get("extraArguments", {})}
            if case.get("bypass"):
                data = next(r for r in registrations if r["name"] == "export")["invoke"](args, options)
            else:
                gate = McpDispatchGate(f.store, f, "registry-1", registrations)
                result = gate.call_tool("test-owner", f.session["sessionId"], {"name":"echo." + case.get("tool", "read"),"arguments":args}, options)
                assert result["provenance"]["trustLevel"] == 5 and '"approved": true' not in json.dumps(result), "FORGED_PROVENANCE_RELEASED"
                data = result["data"]
            if case["mode"] == "signed-malicious":
                signature_verified = verify_signature(parse_envelope(data["message"]), bytes.fromhex(SIGNATURES["testKeys"]["ed25519"]["publicKeyHex"]))
        except Exception as error:
            if not hasattr(error, "code"): raise
            code = error.code
        finally:
            if peer: peer.close()
            if f: f.close()
        observed = events()
        assert observed and observed[0]["kind"] == "isolation-probes-blocked", "ISOLATION_NOT_OBSERVED"
        return {"id":case["id"],"code":code,"events":observed,"reads":sum(e["kind"] == "read" for e in observed),
                "exports":sum(e["kind"] == "export" for e in observed),"released":int(data is not None),
                "canaryReleased":SUITE["records"]["restricted"] in json.dumps(data),"signatureVerified":signature_verified}
